ROADMAP.md

```diff
@@ -0,0 +1,579 @@
+gaining adoption in different languages. This is arguably a sign that
+developers accept the concepts, semantics and APIs that libp2p
+introduces. If we stay on track, libp2p may become the de-facto standard
+library for P2P networking.
+
+When that happens, it'll make sense to push our constructions down one level
+into OS kernels. This can entail first incubating external modules, to
+eventually propose merging them into the main codebases of kernels like Linux.
+An intermediary step would be to run libp2p as a daemon, providing peer-to-peer
+network capabilities to all processes on a machine.
+
+**Why?** To ship libp2p as the default P2P network API inside future
+operating systems. Actually, taking it one step further, make the libp2p
+stack *THE default* networking stack.
+
+### ♻️ Packet switching
+
+**What?** A packet-switched libp2p is on the pathway to achieving
+decentralisation and seamless connectivity across heterogeneous networks
+across the globe, that are running on different transports and not
+necessarily connected to the Internet directly. Essentially it's a
+*global transit system* for libp2p traffic.
+
+For illustration, a TCP packet sent via an IP-based local LAN in
+the US connected to the Internet backbone (via a router) could travel
+through the libp2p network, taking multiple hops along the way based on
+routing decisions, and traversing various networks over different
+transports. Eventually it'll reach its destination in Nepal: a node in a
+mesh network, that's in turn connected to a community network with
+intermittent Internet connection via satellite.
+
+To build a reality like this entails doing away with the concept of
+Autonomous Systems (AS), and rethinking the Internet protocols and
+structures from the viewpoint of peer-to-peer: BGP, OSPF, peering,
+transit networks, multihoming, IANA, etc.
+
+**Why?** It's a crucial requirement in our visionary goal of
+decentralising networks. WIP.
+
+**Links:**
+
+-   [On \"DialMe\", NAT traversal, Relay, line switching, packet switching
+    (libp2p protocols)](https://github.com/ipfs/notes/issues/110)
+
+## Evolve
+
+**Our short-term roadmap**.
+
+This is the stuff pushing the existing libp2p stack forward.
+
+### 🤝 Low latency, efficient connection handshake
+
+**High priority for: IPFS**
+
+**What?** Establishing a connection and performing the initial handshake
+should be as cheap and fast as possible. Supporting things like
+*selective* stream opening, *preemption*, *speculative* negotiation,
+*upfront* negotiation, protocol table *pinning*, etc. may enable us to
+achieve lower latencies when establishing connections, and even the
+0-RTT holy grail in some cases. These features are being discussed in
+the *Protocol Select* protocol design.
+
+**Why?** Multistream 1.0 is chatty and naïve. Streams are essential to
+libp2p, and negotiating them is currently inefficient in a number of
+scenarios. Also, bootstrapping a multiplexed connection is currently
```

Comment:
I assume some of this information is in the docs below, but I think it would be good to call out a couple of numbers here.  For example, (this is me making something up), "The p50 to create a bootstrapped multiplex connect between IPFS nodes currently takes Xms from Y round trips.  With improvements to this space, we believe we can reduce this to Ams and B round trips.  This is important because IPFS nodes establish these connections each time connecting to a peer for doing a lookup or initiating a bitswap session.  The p50 of time to first byte of content download for a cold query will reduce from Cms to Dms."

Reply:
Good call. I don't have these numbers (yet). See https://github.com/protocol/web3-dev-team/pull/119 for a qualitative analysis.